chore: Update Client Node Documentation (#266)
alphagov/notifications-tech-docs: source/documentation/client_docs/_node.md

        from_number: "GOVUK" // required string - sender name / phone number
    },
    uri: "https://api.notifications.service.gov.uk/v2/notifications/740e5834-3a29-46b4-9a6f-16142fde533a", // required string
    template: {
        id: "f33517ff-2a88-4f6e-b855-c550268ce08a", // required string - template ID
        version: 3, // required integer
        uri: "https://api.notifications.service.gov.uk/v2/template/f33517ff-2a88-4f6e-b855-c550268ce08a" // required string
    }
};
```

If you are using the [test API key](#test), all your messages come back with a `delivered` status.
If you are using the [test API key](#test), all your messages will come back with a `delivered` status.

All messages sent using the [team and guest list](#team-and-guest-list) or [live](#live) keys appear on your dashboard.
All messages sent using the [team and guest list](#team-and-guest-list) or [live](#live) keys will appear on your dashboard.

#### Error codes

If the request is not successful, the promise fails with an `err`.
If the request is not successful, the client returns an `HTTPError` containing the relevant error code.

> |err.response.data.status_code|err.response.data.errors|How to fix|

Author: reminder: need to change back

> TODO:Check if the variables would be camel case

Author: reminder: check on node library to see camel case situation

```javascript
{
    "id": "740e5834-3a29-46b4-9a6f-16142fde533a", // required string - notification ID
    "reference": "your reference", // optional string - reference you provided when sending the message
    "content": {
        "body": "Hi Amala, your appointment is on 2018-01-01 at 01:00PM", // required string - message content
        "from_number": "GOVUK" // required string - sender name / phone number
    },
    "uri": "https://api.notifications.service.gov.uk/v2/notifications/740e5834-3a29-46b4-9a6f-16142fde533a", // required string
    "template": {
        "id": "f33517ff-2a88-4f6e-b855-c550268ce08a", // required string - template ID
        "version": 3, // required integer
        "uri": "https://api.notifications.service.gov.uk/v2/template/f33517ff-2a88-4f6e-b855-c550268ce08a" // required string
    }
};
```

If you are using the [test API key](#test), all your messages will come back with a `delivered` status.

All messages sent using the [team and guest list](#team-and-guest-list) or [live](#live) keys will appear on your dashboard.

#### Error codes

> If the request is not successful, the client returns an `HTTPError` containing the relevant error code.

Reviewer: This is consistent with the other docs but it’s less correct now.

The client returns a [`Promise`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise). Developers must write a function to handle what happens if that promise fails. That function is passed as an argument to `Promise.catch` or `Promise.then`. In either case the `err` object detailed in the table below will be available in the scope of the function.

TL;DR we can probably leave this line as it is:
```suggestion
If the request is not successful, the promise fails with an `err`.
```

#### Method

```javascript
let options = {
    personalisation: {
        first_name: "Amala",

>         appointment_date: "2018-01-01 at 01:00PM",

Reviewer: Follow GOV.UK Style for [dates](https://www.gov.uk/guidance/style-guide/a-to-z#dates) and [times](https://www.gov.uk/guidance/style-guide/a-to-z#times).
```suggestion
        appointment_date: "1 January 2018 at 1:00pm",
```

        required_documents: ["passport", "utility bill", "other id"]
    },
    reference: "your reference",
    oneClickUnsubscribeURL: "https://example.com/unsubscribe.html?opaque=123456789",

>     emailReplyToId: "ca4fdde7-2a67-4a6c-8393-62aa7245751f" 

Reviewer: Maybe a wider discussion but I don’t think we should include these in the default snippet.

They are going to get in the way of people just trying to copy and paste some code to get up and running, and most developers won’t need to set or change them.

Author: I think this is a wider discussion, I know @CrystalPea was working on improving docs so I am not sure if that has been considered already

Author: she won't be around for a while as she is on A/L, but if you'd like this to discussed further we can create a ticket

Reviewer: Yeah, I think let’s make a ticket or note it down in a document somewhere

Author: created this ticket:

https://trello.com/c/Z7K4Hrum/938-review-inclusion-of-optional-fields-in-default-code-snippets

};

notifyClient
  .sendEmail(templateId, emailAddress, {
    personalisation: personalisation,
    reference: reference,
    oneClickUnsubscribeURL: oneClickUnsubscribeURL,
    emailReplyToId: emailReplyToId
  })
  .sendEmail(templateId, emailAddress, options) // Pass options as the third argument (optional)
  .then(response => console.log(response))
  .catch(err => console.error(err))
  .catch(err => console.error(err));
```

The method returns a [promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises). The promise will either:
#### Arguments

- resolve with a `response` if successful

> - fail with an `err` if unsuccessful

Reviewer: This is useful context, specific to Javascript (see [previous comment](https://github.com/alphagov/notifications-tech-docs/pull/266/files#r1922243559))

You can upload the following file types:

- CSV (.csv)

> - Image (.jpeg, .jpg, .png)

Reviewer: ```suggestion
- image (.jpeg, .jpg, .png)
```

> use lower case at the start of the bullet

https://www.gov.uk/guidance/style-guide/a-to-z#bullet-points-steps

- Microsoft Excel Spreadsheet (.xlsx)
- Microsoft Word Document (.doc, .docx)
- PDF (.pdf)

> - Text (.json, .odt, .rtf, .txt)

Reviewer: ```suggestion
- text (.json, .odt, .rtf, .txt)
```

> use lower case at the start of the bullet

https://www.gov.uk/guidance/style-guide/a-to-z#bullet-points-steps

Pass the file object as a value into the `personalisation` argument. For example:

```javascript
let fs = require("fs")

fs.readFile("path/to/document.pdf", function (err, pdfFile) {
  console.log(err)
  notifyClient.sendEmail(templateId, emailAddress, {
    personalisation: {
      first_name: "Amala",

>       appointment_date: "2018-01-01 at 01:00PM",

Reviewer: Follow GOV.UK Style for [dates](https://www.gov.uk/guidance/style-guide/a-to-z#dates) and [times](https://www.gov.uk/guidance/style-guide/a-to-z#times).
```suggestion
      appointment_date: "1 January 2018 at 1:00pm",
```

If Notify cannot add the correct file extension, recipients may not be able to open your file.

```javascript
let fs = require("fs")
fs.readFile("path/to/document.csv", function (err, csvFile) {
  console.log(err)
  notifyClient.sendEmail(templateId, emailAddress, {
    personalisation: {
      first_name: "Amala",

>       appointment_date: "2018-01-01 at 01:00PM",

Reviewer: Follow GOV.UK Style for [dates](https://www.gov.uk/guidance/style-guide/a-to-z#dates) and [times](https://www.gov.uk/guidance/style-guide/a-to-z#times).
```suggestion
      appointment_date: "1 January 2018 at 1:00pm",
```

To let the recipient download the file without confirming their email address, set the `confirmEmailBeforeDownload` option to `false`.

```javascript
let fs = require("fs")

fs.readFile("path/to/document.pdf", function (err, pdfFile) {
  console.log(err)
  notifyClient.sendEmail(templateId, emailAddress, {
    personalisation: {
      first_name: "Amala",

>       appointment_date: "2018-01-01 at 01:00PM",

Reviewer: Follow GOV.UK Style for [dates](https://www.gov.uk/guidance/style-guide/a-to-z#dates) and [times](https://www.gov.uk/guidance/style-guide/a-to-z#times).
```suggestion
      appointment_date: "1 January 2018 at 1:00pm",
```

Files sent before 12 April 2023 had a longer default period of 78 weeks (18 months).

```javascript
let fs = require("fs")

fs.readFile("path/to/document.pdf", function (err, pdfFile) {
  console.log(err)
  notifyClient.sendEmail(templateId, emailAddress, {
    personalisation: {
      first_name: "Amala",

>       appointment_date: "2018-01-01 at 01:00PM",

Reviewer: Follow GOV.UK Style for [dates](https://www.gov.uk/guidance/style-guide/a-to-z#dates) and [times](https://www.gov.uk/guidance/style-guide/a-to-z#times).
```suggestion
      appointment_date: "1 January 2018 at 1:00pm",
```

```javascript
{
  "id": "201b576e-c09b-467b-9dfa-9c3b689ee730",  // required string - notification ID
  "reference": "your reference",  // optional string - reference you provided when sending the message
  "content": {
    "subject": "Your upcoming pigeon registration appointment",  // required string - message subject

>     "body": "Dear Amala\r\n\r\nYour pigeon registration appointment is scheduled for 2018-01-01 at 01:00PM.\r\n\r\n Here is a link to your invitation document:\r\nhttps://documents.service.gov.uk/d/YlxDzgNUQYi1Qg6QxIpptA/th46VnrvRxyVO9div6f7hA?key=R0VDmwJ1YzNYFJysAIjQd9yHn5qKUFg-nXHVe3Ioa3A\r\n\r\nPlease bring the invite with you to the appointment.\r\n\r\nYours,\r\nPigeon Affairs Bureau",  // required string - message content - see that the link to document is embedded in the message content

Reviewer: Follow GOV.UK Style for [dates](https://www.gov.uk/guidance/style-guide/a-to-z#dates) and [times](https://www.gov.uk/guidance/style-guide/a-to-z#times).
```suggestion
    "body": "Dear Amala\r\n\r\nYour pigeon registration appointment is scheduled for 1 January 2018 at 1:00pm.\r\n\r\n Here is a link to your invitation document:\r\nhttps://documents.service.gov.uk/d/YlxDzgNUQYi1Qg6QxIpptA/th46VnrvRxyVO9div6f7hA?key=R0VDmwJ1YzNYFJysAIjQd9yHn5qKUFg-nXHVe3Ioa3A\r\n\r\nPlease bring the invite with you to the appointment.\r\n\r\nYours,\r\nPigeon Affairs Bureau",  // required string - message content - see that the link to document is embedded in the message content
```

```javascript
{
    "id": "3d1ce039-5476-414c-99b2-fac1e6add62c",  // required string - notification ID
    "reference": "your reference",  // optional string - reference you provided when sending the message
    "content": {
        "subject": "Your upcoming pigeon registration appointment",  // required string - letter heading

>         "body": "Dear Amala\r\n\r\nYour pigeon registration appointment is scheduled for 2018-01-01 at 01:00PM.\r\n\r\nPlease bring:\r\n\n\n* passport\n* utility bill\n* other id\r\n\r\nYours,\r\nPigeon Affairs Bureau",  // required string - letter content

Reviewer: Follow GOV.UK Style for [dates](https://www.gov.uk/guidance/style-guide/a-to-z#dates) and [times](https://www.gov.uk/guidance/style-guide/a-to-z#times).
```suggestion
        "body": "Dear Amala\r\n\r\nYour pigeon registration appointment is scheduled for 1 January 2018 at 1:00pm.\r\n\r\nPlease bring:\r\n\n\n* passport\n* utility bill\n* other id\r\n\r\nYours,\r\nPigeon Affairs Bureau",  // required string - letter content
```

    "reference": "your reference",  // optional string - reference you provided when sending the message
    "email_address": "amala@example.com",  // required string for emails
    "phone_number": "+447700900123",  // required string for text messages
    "line_1": "Amala Bird",  // required string for letter
    "line_2": "123 High Street",  // required string for letter
    "line_3": "Richmond upon Thames",  // required string for letter
    "line_4": "Middlesex",  // optional string for letter
    "line_5": "SW14 6BF",  // optional string for letter
    "line_6": null,  // optional string for letter
    "line_7": null, // optional string for letter
    "postage": "first / second / europe / rest-of-world", // required string for letter
    "type": "sms / letter / email",  // required string
    "status": "sending / delivered / permanent-failure / temporary-failure / technical-failure",  // required string
    "template": {
        "version": 1, // required integer
        "id": "f33517ff-2a88-4f6e-b855-c550268ce08a",  // required string - template ID
        "uri": "/v2/template/{id}/{version}"  // required string
    },

>     "body": "Hi Amala, your appointment is on 2018-01-01 at 01:00PM",  // required string - body of notification

Reviewer: Follow GOV.UK Style for [dates](https://www.gov.uk/guidance/style-guide/a-to-z#dates) and [times](https://www.gov.uk/guidance/style-guide/a-to-z#times).
```suggestion
    "body": "Hi Amala, your appointment is on 1 January 2018 at 01:00pm",  // required string - body of notification
```

    "subject": "Your upcoming pigeon registration appointment",  // required string for email - subject of email

>     "created_at": "2024-05-17 15:58:38.342838",  // required string - date and time notification created

Reviewer: Do we mention anywhere what timezone this is in?

Author: nope

Reviewer: Maybe this one should also be a ticket so we can review it across all the docs?

    "created_by_name": "Charlie Smith",  // optional string - name of the person who sent the notification if sent manually
    "sent_at": "2024-05-17 15:58:30.143000",  // optional string - date and time notification sent to provider
    "completed_at": "2024-05-17 15:59:10.321000",  // optional string - date and time notification delivered or failed

>     "scheduled_for": "2024-05-17 9:00:00.000000", // optional string - date and time notification has been scheduled to be sent at

Reviewer: Will this ever have a non-`null` value? Should we mark it as deprecated?

Author: do you mean the scheduled_for field?

we do populate it

Reviewer: Yes. It is always `None`(`null` in Javascript):
https://github.com/alphagov/notifications-api/blob/34584cb9529c634acdae2f671a1701a00dec7850/app/v2/notifications/create_response.py#L57

It’s been that way since https://github.com/alphagov/notifications-api/pull/2897

Author: since it is always null I have removed it from the response code snippets

|#`created`|GOV.UK Notify has placed the message in a queue, ready to be sent to the provider. It should only remain in this state for a few seconds.|
|#`sending`|GOV.UK Notify has sent the message to the provider. The provider will try to deliver the message to the recipient for up to 72 hours. GOV.UK Notify is waiting for delivery information.|
|#`pending`|GOV.UK Notify is waiting for more delivery information.<br>GOV.UK Notify received a callback from the provider but the recipient’s device has not yet responded. Another callback from the provider determines the final status of the text message.|

> |#`sent`|The message was sent to an international number. The mobile networks in some countries do not provide any more delivery information. The GOV.UK Notify website displays this status as "Sent to an international number".|

Reviewer: ```suggestion
|#`sent`|The message was sent to an international number. The mobile networks in some countries do not provide any more delivery information. The GOV.UK Notify website displays this status as ‘Sent to an international number’.|
```

> use single quotes when referring to words

– https://www.gov.uk/guidance/style-guide/a-to-z#single-quotes

const notifyClient = new NotifyClient(baseUrl, serviceId, apiKeyId);

// To get the latest 250 received text messages
notifyClient.getReceivedTexts()
    .then((response) => {
        console.log(response); // process received text messages
    })
    .catch((error) => {
        console.error(error);
    });

// To get the next 250 received text messages older than a given reference
const olderThanId = '35836a9e-5a97-4d99-8309-0c5a2c3dbc72';
notifyClient.getReceivedTexts(olderThanId)
    .then((response) => {
        console.log(response); // process received text messages
    })
    .catch((error) => {
        console.error(error);
    });
```

#### Response

If the request to the client is successful, the client will return a `Promise` that resolves to a JSON object containing the received text messages.
```javascript
{
  "status": 200,
  "data": {
    "received_text_messages": [
      {
        "id": "123456789",
        "content": "Sample text message content",
        "created_at": "2024-01-01T12:00:00.000Z",
        "phone_number": "+447700900000",
        // additional message details
      },
      // additional text messages
    ]
  }
}
```

> 

Reviewer: This duplicates the existing content below.

Unlike the Python client the Node client doesn’t have a method way to automatically iterate over all received messages, the developer needs to write a loop themselves.

Author: so should I remove this section?
the main idea with this PR is to try and make all docs uniform, that's why I added this duplication but it's ok to leave it out.

Reviewer: Yes, remove. It only applies to the Python docs because the Python client has an additional method which the other clients do not.

Some discussion of the feature in Python here: https://gds.slack.com/archives/C03S5QFTWSD/p1733419619173219

```javascript
{
    "received_text_messages":
    [
        {

>             "id": "'b51f638b-4295-46e0-a06e-cd41eee7c33b", // required string - ID of received text message

Reviewer: ```suggestion
            "id": "b51f638b-4295-46e0-a06e-cd41eee7c33b", // required string - ID of received text message
```
